File: algo/KthSmallestElementinaBST.java
import java.util.ArrayDeque;
import java.util.PriorityQueue;
import java.util.Queue;

public class KthSmallestElementinaBST {
    public int kthSmallest(TreeNode root, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        Queue<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            pq.add(node.val);
            if (pq.size() > k) {
                pq.remove();
            }
            if (node.left != null) {
                q.add(node.left);
            }
            if (node.right != null) {
                q.add(node.right);
            }
        }
        return pq.peek();
    }
}

File: algo/KthSmallestElementinaBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        dfs(root, k);
        return list[k-1];
    }
private:
    vector<int> list;
    void dfs(TreeNode* node, int k) {
        if (node == NULL) {
            return;
        }
        if (list.size() < k) {
            dfs(node->left, k);
            list.push_back(node->val);
            dfs(node->right, k);
        } else {
            return;
        }
    }
};